roofit/roofitcore/test/TestStatistics/RooRealL.cpp

```diff
@@ -0,0 +1,338 @@
+// Author: Patrick Bos, Inti Pelupessy, Vince Croft, Netherlands eScience Center / NIKHEF 2021
+
+/*****************************************************************************
+ * RooFit
+ * Authors:                                                                  *
+ *   WV, Wouter Verkerke, UC Santa Barbara, verkerke@slac.stanford.edu       *
+ *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
+ *                                                                           *
+ * Copyright (c) 2000-2021, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+#include "TestStatistics/RooRealL.h"
+#include "TestStatistics/RooUnbinnedL.h"
+
+#include <RooRandom.h>
+#include <RooWorkspace.h>
+#include <RooAbsPdf.h>
+#include <RooDataSet.h>
+#include <RooMinimizer.h>
+#include <RooFitResult.h>
+#include <RooProdPdf.h>
+#include <RooAddition.h>
+#include <RooConstraintSum.h>
+#include <RooPolynomial.h>
+#include <RooDataHist.h>
+#include <RooRealSumPdf.h>
+#include <RooNLLVar.h>
+
+#include "gtest/gtest.h"
+
+class RooRealL
+   : public ::testing::TestWithParam<std::tuple<std::size_t>> {
+};
+
+TEST_P(RooRealL, getVal)
+{
+   // Real-life test: calculate a NLL using event-based parallelization. This
+   // should replicate RooRealMPFE results.
+   RooRandom::randomGenerator()->SetSeed(std::get<0>(GetParam()));
+   RooWorkspace w;
+   w.factory("Gaussian::g(x[-5,5],mu[0,-3,3],sigma[1])");
+   auto x = w.var("x");
+   RooAbsPdf *pdf = w.pdf("g");
+   RooDataSet *data = pdf->generate(RooArgSet(*x), 10000);
+   auto nll = pdf->createNLL(*data);
+
+   auto nominal_result = nll->getVal();
+
+   RooFit::TestStatistics::RooRealL nll_new("nll_new", "new style NLL", std::make_shared<RooFit::TestStatistics::RooUnbinnedL>(pdf, data));
+
+   auto mp_result = nll_new.getVal();
+
+   EXPECT_DOUBLE_EQ(nominal_result, mp_result);
+}
+
+void check_NLL_type(RooAbsReal *nll)
+{
+   if (dynamic_cast<RooAddition *>(nll) != nullptr) {
+      std::cout << "the NLL object is a RooAddition*..." << std::endl;
+      bool has_rooconstraintsum = false;
+      RooFIter nll_component_iter = nll->getComponents()->fwdIterator();
+      RooAbsArg *nll_component;
+      while ((nll_component = nll_component_iter.next())) {
```

Comment:
It's "only" a test, but the range-base for loop is faster and cleaner

Reply:
Ok, since it's only a test, I went ahead and replaced the other RooFIter + for loop below with `std::count_if` and a generic lambda expression 😄 C++14! 🎉

Reply:
I really hate that those `std` algo's still need begin and end, but I guess I'll survive.